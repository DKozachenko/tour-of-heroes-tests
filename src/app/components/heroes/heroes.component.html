<h2 automation-id="heading" pw-automation-id="heading">My Heroes</h2>

<div automation-id="create-div" pw-automation-id="create-div">
  <label
    for="new-hero"
    automation-id="create-label"
    pw-automation-id="create-label"
    >Hero name:
  </label>
  <input
    id="new-hero"
    #heroName
    automation-id="create-input"
    pw-automation-id="create-input"
  />

  <!-- (click) passes input value to add() and then clears the input -->
  <button
    type="button"
    class="add-button"
    (click)="add(heroName.value); heroName.value = ''"
    automation-id="add-button"
    pw-automation-id="add-button"
  >
    Add hero
  </button>
</div>

<ul class="heroes" automation-id="heroes" pw-automation-id="heroes">
  <li
    *ngFor="let hero of heroes"
    automation-id="hero-item"
    pw-automation-id="hero-item"
    cy-automation-id="hero-item"
  >
    <a
      routerLink="/detail/{{ hero.id }}"
      automation-id="hero-link"
      pw-automation-id="hero-link"
    >
      <span class="badge">{{ hero.id }}</span> {{ hero.name }}
    </a>
    <button
      type="button"
      class="delete"
      title="delete hero"
      (click)="delete(hero)"
      automation-id="hero-button"
      pw-automation-id="hero-button"
    >
      x
    </button>
  </li>
</ul>
