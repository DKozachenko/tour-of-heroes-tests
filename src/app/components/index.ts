export { DashboardComponent } from './dashboard/dashboard.component';
export { HeroDetailComponent } from './hero-detail/hero-detail.component';
export { HeroSearchComponent } from './hero-search/hero-search.component';
export { HeroesComponent } from './heroes/heroes.component';
export { MessagesComponent } from './messages/messages.component';
export { AppComponent } from './app/app.component';
