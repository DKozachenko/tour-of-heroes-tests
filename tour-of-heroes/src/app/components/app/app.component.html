<h1 automation-id="heading">Tour of Heroes</h1>
<nav automation-id="navigation">
  <a routerLink="/dashboard" automation-id="navigation-link">Dashboard</a>
  <a routerLink="/heroes" automation-id="navigation-link">Heroes</a>
</nav>
<router-outlet></router-outlet>
<app-messages automation-id="messages"></app-messages>
