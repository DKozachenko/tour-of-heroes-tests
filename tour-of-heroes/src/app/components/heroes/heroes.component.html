<h2 automation-id="heading">My Heroes</h2>

<div automation-id="create-div">
  <label for="new-hero" automation-id="create-label">Hero name: </label>
  <input id="new-hero" #heroName automation-id="create-input" />

  <!-- (click) passes input value to add() and then clears the input -->
  <button
    type="button"
    class="add-button"
    (click)="add(heroName.value); heroName.value = ''"
    automation-id="add-button"
  >
    Add hero
  </button>
</div>

<ul class="heroes" automation-id="heroes">
  <li *ngFor="let hero of heroes" automation-id="hero-item">
    <a routerLink="/detail/{{ hero.id }}" automation-id="hero-link">
      <span class="badge">{{ hero.id }}</span> {{ hero.name }}
    </a>
    <button
      type="button"
      class="delete"
      title="delete hero"
      (click)="delete(hero)"
      automation-id="hero-button"
    >
      x
    </button>
  </li>
</ul>
