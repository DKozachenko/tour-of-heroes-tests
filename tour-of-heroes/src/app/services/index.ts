export { HeroService } from './hero/hero.service';
export {
  InMemoryDataService,
  HEROES,
} from './in-memory-data/in-memory-data.service';
export { MessageService } from './message/message.service';
