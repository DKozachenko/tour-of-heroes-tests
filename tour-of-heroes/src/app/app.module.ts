import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemoryDataService } from './services';

import {
  DashboardComponent,
  HeroDetailComponent,
  HeroesComponent,
  HeroSearchComponent,
  MessagesComponent,
} from './components';

@NgModule({
  imports: [
    BrowserModule,
    FormsModule,
    AppRoutingModule,
    HttpClientModule,

    // The HttpClientInMemoryWebApiModule module intercepts HTTP requests
    // and returns simulated server responses.
    // Remove it when a real server is ready to receive requests.
    HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, {
      dataEncapsulation: false,
    }),
  ],
  declarations: [
    AppComponent,
    DashboardComponent,
    HeroesComponent,
    HeroDetailComponent,
    MessagesComponent,
    HeroSearchComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
