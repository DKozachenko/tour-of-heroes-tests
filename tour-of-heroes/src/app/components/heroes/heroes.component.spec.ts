import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { MockBuilder, MockRender, MockedComponentFixture } from 'ng-mocks';
import {
  mock,
  instance,
  when,
  verify,
  anything,
  anyOfClass,
  anyString,
  deepEqual,
} from 'ts-mockito';
import { of } from 'rxjs';
import { HEROES, HeroService } from '../../services';
import { Hero } from '../../models';
import { HeroesComponent } from './heroes.component';
import { AppModule } from '../../app.module';
import { RouterLink } from '@angular/router';

class PageObject {
  private fixtureDebugElement: DebugElement;

  constructor(fixture: MockedComponentFixture<HeroesComponent>) {
    this.fixtureDebugElement = fixture.debugElement;
  }

  get heading(): DebugElement {
    return this.fixtureDebugElement.query(By.css('h2'));
  }

  get createDiv(): DebugElement {
    return this.fixtureDebugElement.query(By.css('div'));
  }

  get label(): DebugElement {
    return this.createDiv.query(By.css('label'));
  }

  get input(): DebugElement {
    return this.createDiv.query(By.css('input'));
  }

  get addButton(): DebugElement {
    return this.createDiv.query(By.css('button'));
  }

  get heroList(): DebugElement {
    return this.fixtureDebugElement.query(By.css('.heroes'));
  }

  get heroItems(): DebugElement[] {
    return this.heroList.queryAll(By.css('li'));
  }

  get heroItemsLinks(): DebugElement[] {
    return this.heroList.queryAll(By.css('li a'));
  }

  get heroItemsButtons(): DebugElement[] {
    return this.heroList.queryAll(By.css('li button'));
  }
}

describe('HeroesComponent', () => {
  let mockHeroService: HeroService;

  beforeEach(() => {
    mockHeroService = mock(HeroService);

    return MockBuilder(HeroesComponent, AppModule).mock(
      HeroService,
      instance(mockHeroService)
    );
  });

  function createFixture(): MockedComponentFixture<HeroesComponent> {
    return MockRender(HeroesComponent);
  }

  function createComponent(): HeroesComponent {
    return createFixture().point.componentInstance;
  }

  function mockCalls(): void {
    when(mockHeroService.getHeroes()).thenReturn(of(HEROES));
  }

  // Functional tests
  it('should call "getHeroes" service method while creating', () => {
    mockCalls();
    createComponent();
    verify(mockHeroService.getHeroes()).once();
  });

  it('should call "getHeroes" service method via "getHeroes" component method', () => {
    mockCalls();
    const component = createComponent();
    component.getHeroes();
    verify(mockHeroService.getHeroes()).twice();
  });

  it('should not call "addHero" service method via "add" component method if name is empty string (or contains only spaces)', () => {
    mockCalls();
    const component = createComponent();

    const spyOnServiceAddHero = jest.spyOn(mockHeroService, 'addHero');
    component.add('');
    component.add('      ');
    expect(spyOnServiceAddHero).not.toHaveBeenCalled();
  });

  it('should call "addHero" service method via "add" component method and push new hero in "heroes" property', () => {
    const oldHeroes = [HEROES[0], HEROES[1]];
    const newHeroName = ' test name ';
    const mockNewHero: Hero = {
      id: oldHeroes[oldHeroes.length - 1].id + 1,
      name: newHeroName.trim(),
    };

    const spyOnServiceAddHero = jest.spyOn(mockHeroService, 'addHero');
    when(mockHeroService.getHeroes()).thenReturn(of([...oldHeroes]));
    when(
      mockHeroService.addHero(deepEqual({ name: mockNewHero.name } as Hero))
    ).thenReturn(of(mockNewHero));

    const component = createComponent();
    component.add(newHeroName);
    // TODO: можно ли без deepEqual
    expect(spyOnServiceAddHero).toHaveBeenCalledWith(
      deepEqual({ name: mockNewHero.name })
    );
    expect(spyOnServiceAddHero).toHaveBeenCalled();

    expect(component.heroes.length).toBe(oldHeroes.length + 1);
    expect(component.heroes).toStrictEqual(oldHeroes.concat([mockNewHero]));
  });

  it('should not change "heroes" property if non-existent hero was tried to deleted', () => {
    const oldHeroes = [HEROES[0], HEROES[1]];
    const nonExistentHero = {
      id: 1000,
      name: 'non-existent',
    };
    const spyOnServiceDeleteHero = jest.spyOn(mockHeroService, 'deleteHero');
    when(mockHeroService.getHeroes()).thenReturn(of(oldHeroes));
    when(mockHeroService.deleteHero(nonExistentHero.id)).thenReturn(
      of(undefined as any)
    );

    const component = createComponent();
    component.delete(nonExistentHero);
    expect(component.heroes.length).toBe(oldHeroes.length);
    expect(component.heroes).toStrictEqual(oldHeroes);
    expect(spyOnServiceDeleteHero).toHaveBeenCalledWith(nonExistentHero.id);
  });

  it('should change "heroes" property if existed hero was tried to deleted', () => {
    const oldHeroes = [HEROES[0], HEROES[1]];
    const nonExistentHero = oldHeroes[0];
    const spyOnServiceDeleteHero = jest.spyOn(mockHeroService, 'deleteHero');
    when(mockHeroService.getHeroes()).thenReturn(of(oldHeroes));
    when(mockHeroService.deleteHero(nonExistentHero.id)).thenReturn(
      of(undefined as any)
    );

    const component = createComponent();
    component.delete(nonExistentHero);
    expect(component.heroes.length).toBe(oldHeroes.length - 1);
    expect(component.heroes).toStrictEqual(
      oldHeroes.filter((hero) => hero.id !== nonExistentHero.id)
    );
    expect(spyOnServiceDeleteHero).toHaveBeenCalledWith(nonExistentHero.id);
  });

  // Layout tests
  it('should contain heading', () => {
    mockCalls();

    const fixture = createFixture();
    const pageObject = new PageObject(fixture);
    expect(pageObject.heading).not.toBeNull();
  });

  it('should contain block for creating new hero with input and add button', () => {
    mockCalls();

    const fixture = createFixture();
    const pageObject = new PageObject(fixture);
    expect(pageObject.label).not.toBeNull();
    expect(pageObject.label.nativeElement.textContent).toBe('Hero name: ');
    expect(pageObject.input).not.toBeNull();
    expect(pageObject.addButton).not.toBeNull();
    expect(pageObject.addButton.nativeElement.textContent).toContain(
      'Add hero'
    );
  });

  it('should call "add" component method and clear input if add button has clicked', () => {
    mockCalls();
    when(mockHeroService.addHero(anything())).thenReturn(of(anything()));

    const fixture = createFixture();
    const component = fixture.point.componentInstance;
    const spyOnAdd = jest.spyOn(component, 'add');

    const pageObject = new PageObject(fixture);
    const mockNewHeroName = 'test name';
    pageObject.input.nativeElement.value = mockNewHeroName;
    pageObject.addButton.triggerEventHandler('click');
    expect(spyOnAdd).toHaveBeenCalledWith(mockNewHeroName);
    expect(pageObject.input.nativeElement.value).toBe('');
  });

  it('should container hero list with link and button for deleting', () => {
    const mockHeroes = [HEROES[0], HEROES[1]];
    when(mockHeroService.getHeroes()).thenReturn(of(mockHeroes));

    const fixture = createFixture();
    const pageObject = new PageObject(fixture);
    expect(pageObject.heroList).not.toBeNull();
    expect(pageObject.heroItems.length).toBe(mockHeroes.length);
    for (let i = 0; i < pageObject.heroItems.length; ++i) {
      const hero = mockHeroes[i];
      const heroItemLink = pageObject.heroItemsLinks[i];
      const heroItemButton = pageObject.heroItemsButtons[i];
      expect(heroItemLink.nativeElement.textContent).toContain(
        `${hero.id} ${hero.name}`
      );
      expect(heroItemLink.injector.get(RouterLink).routerLink).toBe(
        `/detail/${hero.id}`
      );
      expect(heroItemButton.nativeElement.textContent).toContain('x');
    }
  });

  it('should add new item to hero list if add button has clicked', () => {
    const oldHeroes = [HEROES[0], HEROES[1]];
    const mockNewHero: Hero = {
      id: oldHeroes[oldHeroes.length - 1].id + 1,
      name: 'new hero name',
    };

    when(mockHeroService.getHeroes()).thenReturn(of([...oldHeroes]));
    when(
      mockHeroService.addHero(deepEqual({ name: mockNewHero.name } as Hero))
    ).thenReturn(of(mockNewHero));

    const fixture = createFixture();
    const pageObject = new PageObject(fixture);
    expect(pageObject.heroList).not.toBeNull();
    expect(pageObject.heroItems.length).toBe(oldHeroes.length);
    for (let i = 0; i < pageObject.heroItems.length; ++i) {
      const hero = oldHeroes[i];
      const heroItemLink = pageObject.heroItemsLinks[i];
      const heroItemButton = pageObject.heroItemsButtons[i];
      expect(heroItemLink.nativeElement.textContent).toContain(
        `${hero.id} ${hero.name}`
      );
      expect(heroItemLink.injector.get(RouterLink).routerLink).toBe(
        `/detail/${hero.id}`
      );
      expect(heroItemButton.nativeElement.textContent).toContain('x');
    }

    pageObject.input.nativeElement.value = mockNewHero.name;
    pageObject.addButton.triggerEventHandler('click');
    fixture.detectChanges();
    const newHeroes = oldHeroes.concat([mockNewHero]);
    expect(pageObject.heroItems.length).toBe(newHeroes.length);
    for (let i = 0; i < pageObject.heroItems.length; ++i) {
      const hero = newHeroes[i];
      const heroItemLink = pageObject.heroItemsLinks[i];
      const heroItemButton = pageObject.heroItemsButtons[i];
      expect(heroItemLink.nativeElement.textContent).toContain(
        `${hero.id} ${hero.name}`
      );
      expect(heroItemLink.injector.get(RouterLink).routerLink).toBe(
        `/detail/${hero.id}`
      );
      expect(heroItemButton.nativeElement.textContent).toContain('x');
    }
  });

  it('should call "delete" component method if remove item button has clicked', () => {
    const mockHeroes = [HEROES[0], HEROES[1], HEROES[2]];
    const heroIndexForRemove = 1;
    const heroForRemove = mockHeroes[heroIndexForRemove];
    when(mockHeroService.getHeroes()).thenReturn(of(mockHeroes));
    when(mockHeroService.deleteHero(heroForRemove.id)).thenReturn(
      of(undefined as any)
    );

    const fixture = createFixture();
    const component = fixture.point.componentInstance;
    const spyOnAdd = jest.spyOn(component, 'delete');

    const pageObject = new PageObject(fixture);
    pageObject.heroItemsButtons[heroIndexForRemove].triggerEventHandler(
      'click'
    );
    expect(spyOnAdd).toHaveBeenCalledWith(heroForRemove);
  });

  it('should remove item from hero list if remove item button has clicked', () => {
    const oldHeroes = [HEROES[0], HEROES[1], HEROES[2]];
    const heroIndexForRemove = 1;
    const heroForRemove = oldHeroes[heroIndexForRemove];
    when(mockHeroService.getHeroes()).thenReturn(of(oldHeroes));
    when(mockHeroService.deleteHero(heroForRemove.id)).thenReturn(
      of(undefined as any)
    );

    const fixture = createFixture();
    const pageObject = new PageObject(fixture);
    expect(pageObject.heroList).not.toBeNull();
    expect(pageObject.heroItems.length).toBe(oldHeroes.length);
    for (let i = 0; i < pageObject.heroItems.length; ++i) {
      const hero = oldHeroes[i];
      const heroItemLink = pageObject.heroItemsLinks[i];
      const heroItemButton = pageObject.heroItemsButtons[i];
      expect(heroItemLink.nativeElement.textContent).toContain(
        `${hero.id} ${hero.name}`
      );
      expect(heroItemLink.injector.get(RouterLink).routerLink).toBe(
        `/detail/${hero.id}`
      );
      expect(heroItemButton.nativeElement.textContent).toContain('x');
    }

    pageObject.heroItemsButtons[heroIndexForRemove].triggerEventHandler(
      'click'
    );
    fixture.detectChanges();
    const newHeroes = oldHeroes.filter((hero) => hero.id !== heroForRemove.id);
    expect(pageObject.heroItems.length).toBe(newHeroes.length);
    for (let i = 0; i < pageObject.heroItems.length; ++i) {
      const hero = newHeroes[i];
      const heroItemLink = pageObject.heroItemsLinks[i];
      const heroItemButton = pageObject.heroItemsButtons[i];
      expect(heroItemLink.nativeElement.textContent).toContain(
        `${hero.id} ${hero.name}`
      );
      expect(heroItemLink.injector.get(RouterLink).routerLink).toBe(
        `/detail/${hero.id}`
      );
      expect(heroItemButton.nativeElement.textContent).toContain('x');
    }
  });
});
