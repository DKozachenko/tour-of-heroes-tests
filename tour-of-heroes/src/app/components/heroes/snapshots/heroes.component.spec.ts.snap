// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesComponent Layout hero list should contain hero list with link and button for deleting 1`] = `
<ul
  automation-id="heroes"
  class="heroes"
>
  <li
    automation-id="hero-item"
    pw-automation-id="hero-item"
  >
    <a
      automation-id="hero-link"
    >
      <span
        class="badge"
      >
        12
      </span>
       Dr. Nice 
    </a>
    
    <button
      automation-id="hero-button"
      class="delete"
      title="delete hero"
      type="button"
    >
       x 
    </button>
  </li>
  <li
    automation-id="hero-item"
    pw-automation-id="hero-item"
  >
    <a
      automation-id="hero-link"
    >
      <span
        class="badge"
      >
        13
      </span>
       Bombasto 
    </a>
    
    <button
      automation-id="hero-button"
      class="delete"
      title="delete hero"
      type="button"
    >
       x 
    </button>
  </li>
  
</ul>
`;

exports[`HeroesComponent Layout hero list should contain hero list with link and button for deleting 2`] = `
<a
  automation-id="hero-link"
>
  <span
    class="badge"
  >
    12
  </span>
   Dr. Nice 
</a>
`;

exports[`HeroesComponent Layout hero list should contain hero list with link and button for deleting 3`] = `
<button
  automation-id="hero-button"
  class="delete"
  title="delete hero"
  type="button"
>
   x 
</button>
`;

exports[`HeroesComponent Layout hero list should contain hero list with link and button for deleting 4`] = `
<a
  automation-id="hero-link"
>
  <span
    class="badge"
  >
    13
  </span>
   Bombasto 
</a>
`;

exports[`HeroesComponent Layout hero list should contain hero list with link and button for deleting 5`] = `
<button
  automation-id="hero-button"
  class="delete"
  title="delete hero"
  type="button"
>
   x 
</button>
`;

exports[`HeroesComponent Layout should contain block for creating new hero with input and add button 1`] = `
<label
  automation-id="create-label"
  for="new-hero"
>
  Hero name: 
</label>
`;

exports[`HeroesComponent Layout should contain block for creating new hero with input and add button 2`] = `
<input
  automation-id="create-input"
  id="new-hero"
/>
`;

exports[`HeroesComponent Layout should contain block for creating new hero with input and add button 3`] = `
<button
  automation-id="add-button"
  class="add-button"
  type="button"
>
   Add hero 
</button>
`;

exports[`HeroesComponent Layout should contain heading 1`] = `
<h2
  automation-id="heading"
>
  My Heroes
</h2>
`;
